std/string.d

```diff
@@ -357,10 +357,21 @@ alias CaseSensitive = Flag!"caseSensitive";
 
     See_Also: $(REF countUntil, std,algorithm,searching)
   +/
-ptrdiff_t indexOf(Range)(Range s, in dchar c,
-        in CaseSensitive cs = Yes.caseSensitive)
-if (isInputRange!Range && isSomeChar!(ElementEncodingType!Range) &&
-    !isConvertibleToString!Range)
+ptrdiff_t indexOf(Range)(Range s, dchar c, CaseSensitive cs = Yes.caseSensitive)
+if (isInputRange!Range && isSomeChar!(ElementType!Range) && !isSomeString!Range)
+{
+    return _indexOf(s, c, cs);
+}
+
+/// Ditto
+ptrdiff_t indexOf(C)(C[] s, dchar c, CaseSensitive cs = Yes.caseSensitive)
+if (isSomeChar!C)
+{
+    return _indexOf(s, c, cs);
+}
+
+private ptrdiff_t _indexOf(Range)(Range s, dchar c, CaseSensitive cs = Yes.caseSensitive)
```

Comment:
One future PR would do good to unify these overloads, put this implementation inside, etc...

Reply:
That's not possible due to D's implicit conversion from static arrays to dynamic arrays. See @jmdavis's explanation here or https://github.com/dlang/phobos/pull/5410#discussion_r164238570

Reply:
Yeah, I'm of the opinion that we should deprecate implicitly converting static arrays to dynamic arrays (require explicit slicing, e.g., `staticArr[]` instead of just `staticArr`). That is an important first step towards making it possible to merge these overloads.

Reply:
auto ref could help that situash

Reply:
@andralex: I think you misunderstood something. The problem is that `indexOf` was originally declared to accept arrays, something like this:
````
ptrdiff_t indexOf(C)(C[] str, dchar c, CaseSensitive cs = Yes.caseSensitive) ...
````
This original function accepted dynamic arrays, obviously, *but* due to static arrays implicitly converting to dynamic arrays, it also accepted static arrays as well:
````
string x; indexOf(x, 'a'); // OK
char[3] y; indexOf(y, 'b'); // also OK (implicit conversion of char[3] to char[])
````

Then we generalized it to take an input range instead:
````
ptrdiff_t indexOf(R)(R r, dchar c, CaseSensitive cs = Yes.caseSensitive)
if (isInputRange!R) ...
````

Unfortunately, static arrays are *not* input ranges (and rightly so, because you can't shrink their length so `popFront` cannot be implemented for them). So now we have the situation:
````
string x; indexOf(x, 'a'); // OK
char[3] y; indexOf(y, 'b'); // Compile error: breaks existing code
````
`auto ref` has nothing to do with this; the underlying problem is that the implicit conversion of static arrays to dynamic arrays only happens if the function parameter is explicitly declared as an array; if it's a generic type `R` with a sig constraint, this conversion never happens, so `R` winds up being `char[3]`, which is not an input range.  Adding `auto ref` *at most* only turns that into `ref char[3]`, which is *still* not an input range.

Furthermore, this is not fixable by shunting the array overload into a private overload set, e.g., I thought of doing this:
````
ptrdiff_t indexOf(R)(R r, dchar ch)
// accept everything here
{
    return _indexOfImpl(r, ch);
}

ptrdiff_t _indexOfImpl(R r, dchar ch)
if (isInputRange!R && !isSomeString!R) // exclude the char[3] case
{...}

ptrdiff_t _indexOfImpl(C)(C[] arr, dchar ch) // induce implicit conversion char[3] -> char[] here
{ ... }
````
Unfortunately, since static arrays are passed by value, this means that if the function returns a modified range, then `_indexOfImpl` would be slicing a local variable (function parameter of `indexOf`) instead of having the slice taken *in the caller's scope*. This would violate `@safe`.

Basically, it all boils down to static arrays being implicitly sliced into dynamic arrays causing problems with generalizing functions that take dynamic arrays to functions that take ranges. Not to mention, such implicit conversions are a bad idea to begin with, because you can inadvertently return an escaping reference to a local static array due to `T[n]` decaying into `T[]` autoomatically. Had we required explicit slicing on the caller's side, i.e., `indexOf(staticArr[])` instead of `indexOf(staticArr)`, none of this would have been a problem.

Reply:
I wasn't misunderstanding. My point is very simple. The following:

```d
ptrdiff_t indexOf(R)(auto ref R r, dchar c, CaseSensitive cs = Yes.caseSensitive)
if (appropriately_chosen_constraints)
{
    static if (zis) { ... }
    else static if (zat) { ... }
    else static if (ze_ozer) { ... }
}
```

covers everything: input ranges, dynamic arrays, static arrays (efficiently, too) with no problem. The only trick is defining the constraint properly.

Reply:
In this particular case, yes, that could be done. Because no reference to the string is returned, but in any of the cases where the string does escape, the implicit conversion must be done at the call site. Regardless, I think that we need to expunge `isConvertibleToString`, because it makes it too easy to get this wrong and can't be used safely in any situation where a reference to the resultant string escapes. Many of the places that it's currently being used in Phobos are not `@safe` when used with something like a static array.